### PR TITLE
feat: report low power mode and thermal info to stats

### PR DESCRIPTION
**Pull request by kristian-mkd**
## Overview

Adds react native implementation for reporting device `thermal state` and `low power mode enabled` states to the call stats.

### CODE REVIEW

File: sample-apps/react-native/dogfood/ios/StreamReactNativeVideoSDKSample/AppDelegate.mm [hunk 1]
```diff
@@ -17,6 +17,10 @@
 
 #import "StreamVideoReactNative.h"
 
+// Thermal module related
```
Comment (kristian-mkd) [inline sample-apps/react-native/dogfood/ios/StreamReactNativeVideoSDKSample/AppDelegate.mm:20]: I will check if these can be removed.

File: packages/react-native-sdk/src/providers/StreamCall.tsx [hunk 2]
```diff
@@ -134,3 +151,161 @@ const ClearPushWSSubscriptions = () => {
...
+  const [thermalState, setThermalState] = useState<string | null>(null);
+
+  const [androidPowerSubscription, setAndroidPowerSubscription] =
+    useState<EmitterSubscription | null>(null);
+  const [androidThermalSubscription, setAndroidThermalSubscription] =
+    useState<EmitterSubscription | null>(null);
+  const [iosPowerSubscription, setIosPowerSubscription] =
+    useState<EmitterSubscription | null>(null);
+  const [iosThermalSubscription, setIosThermalSubscription] =
+    useState<EmitterSubscription | null>(null);
```
Comment (santhoshvai) [inline packages/react-native-sdk/src/providers/StreamCall.tsx:172]: why different event listener names for android and iOS?

File: packages/react-native-sdk/src/providers/StreamCall.tsx [hunk 3]
```diff
@@ -134,3 +151,161 @@ const ClearPushWSSubscriptions = () => {
...
+  if (Platform.OS === 'ios' && !DeviceState) {
+    throw new Error('DeviceState is not properly linked.');
+  }
+
+  let eventEmitter: NativeEventEmitter;
+  if (Platform.OS === 'android') {
+    eventEmitter = new NativeEventEmitter(StreamVideoReactNative);
+  } else {
+    eventEmitter = new NativeEventEmitter(DeviceState);
+  }
```
Comment (santhoshvai) [inline packages/react-native-sdk/src/providers/StreamCall.tsx:188]: Do this outside the render function!

File: packages/react-native-sdk/src/providers/StreamCall.tsx [hunk 4]
```diff
@@ -134,3 +151,161 @@ const ClearPushWSSubscriptions = () => {
...
+    StreamVideoReactNative,
+    StreamVideoReactNative.POWER_MODE_EVENT,
+    StreamVideoReactNative.THERMAL_EVENT,
+    DeviceState,
+    androidPowerSubscription,
+    androidThermalSubscription,
+    iosPowerSubscription,
+    iosThermalSubscription,
+  ]);
+
```
Comment (santhoshvai) [inline packages/react-native-sdk/src/providers/StreamCall.tsx:309]: wow a huge dependency array, for a simple two method implementation
please recheck how you can simplify this

File: packages/react-native-sdk/ios/StreamVideoReactNative-Bridging-Header.h [hunk 5]
```diff
@@ -12,3 +12,4 @@
 #import <WebRTC/RTCVideoRenderer.h>
 #import <WebRTC/RTCVideoFrameBuffer.h>
 #import "WebRTCModule.h"
+#import "ProcessorProvider.h"
```
Comment (santhoshvai) [inline packages/react-native-sdk/ios/StreamVideoReactNative-Bridging-Header.h:15]: why was this added? something didnt work?

File: packages/react-native-sdk/android/src/main/java/com/streamvideo/reactnative/StreamVideoReactNativeModule.kt [hunk 6]
```diff
@@ -81,6 +90,117 @@ class StreamVideoReactNativeModule(reactContext: ReactApplicationContext) : Reac
...
+            if (intent?.action == PowerManager.ACTION_POWER_SAVE_MODE_CHANGED) {
+                sendPowerModeEvent()
+            }
+        }
+    }
+
+    override fun onCatalystInstanceDestroy() {
+        super.onCatalystInstanceDestroy()
+        reactApplicationContext.unregisterReceiver(powerReceiver)
+    }
```
Comment (santhoshvai) [inline packages/react-native-sdk/android/src/main/java/com/streamvideo/reactnative/StreamVideoReactNativeModule.kt:177]: Call stopThermalStatusUpdates here also

File: packages/react-native-sdk/src/providers/StreamCall.tsx [hunk 7]
```diff
@@ -134,3 +151,104 @@ const ClearPushWSSubscriptions = () => {
...
+  const [powerModeSubscription, setPowerModeSubscription] =
+    useState<EmitterSubscription | null>(null);
+  const [thermalStateSubscription, setThermalStateSubscription] =
+    useState<EmitterSubscription | null>(null);
+
+  const { StreamVideoReactNative } = NativeModules;
+  if (!StreamVideoReactNative) {
+    throw new Error('StreamVideoReactNative is not properly linked.');
+  }
```
Comment (santhoshvai) [inline packages/react-native-sdk/src/providers/StreamCall.tsx:173]: move it out of the render function please, its a global constant

thenafter you can remove it from useEffect dependency

File: packages/react-native-sdk/src/providers/StreamCall.tsx [hunk 8]
```diff
@@ -134,3 +151,104 @@ const ClearPushWSSubscriptions = () => {
...
+        isLowPowerMode: lowPowerMode || false,
+      });
+    },
+    [lowPowerMode]
+  );
+
+  useEffect(() => {
+    if (!eventEmitterRef.current) {
+      eventEmitterRef.current = new NativeEventEmitter(StreamVideoReactNative);
+    }
```
Comment (santhoshvai) [inline packages/react-native-sdk/src/providers/StreamCall.tsx:204]: why do this in ref + useEffect? this should be done globally too

File: packages/react-native-sdk/src/providers/StreamCall.tsx [hunk 9]
```diff
@@ -134,3 +151,104 @@ const ClearPushWSSubscriptions = () => {
...
+  const { useCallCallingState } = useCallStateHooks();
+  const callingState = useCallCallingState();
+
+  const [lowPowerMode, setLowPowerMode] = useState<boolean | null>(null);
+  const [thermalState, setThermalState] = useState<string | null>(null);
+
+  const [powerModeSubscription, setPowerModeSubscription] =
+    useState<EmitterSubscription | null>(null);
+  const [thermalStateSubscription, setThermalStateSubscription] =
+    useState<EmitterSubscription | null>(null);
```
Comment (santhoshvai) [inline packages/react-native-sdk/src/providers/StreamCall.tsx:168]: subscriptions shouldnt be needed to be added into a state

get a subscription in useEffect and when returning clear it

File: packages/client/src/client-details.ts [hunk 10]
```diff
@@ -62,6 +68,82 @@ export type LocalClientDetailsType = ClientDetails & {
...
+      android: {
+        thermalState,
+        isPowerSaverMode:
+          deviceState?.oneofKind === 'android' &&
+          deviceState.android.isPowerSaverMode,
+      },
+    };
+  }
+
+  if (osInfo.name === 'iOS') {
```
Comment (oliverlaz) [inline packages/client/src/client-details.ts:93]: I'd do `.toLowerCase()` just to make sure we don't run into weird cases.

File: packages/react-native-sdk/src/providers/StreamCall.tsx [hunk 11]
```diff
@@ -137,3 +153,57 @@ const ClearPushWSSubscriptions = () => {
...
+/**
+ * This is a renderless component to get the device stats like thermal state and power saver mode.
+ */
+const DeviceStats = () => {
+  const { useCallCallingState } = useCallStateHooks();
+  const callingState = useCallCallingState();
+
+  useEffect(() => {
+    let powerModeSubscription: EmitterSubscription;
+    let thermalStateSubscription: EmitterSubscription;
```
Comment (oliverlaz) [inline packages/react-native-sdk/src/providers/StreamCall.tsx:170]: Can you inline these two variable declarations? We don't have if-else anymore.

File: packages/react-native-sdk/src/providers/StreamCall.tsx [hunk 12]
```diff
@@ -137,3 +153,57 @@ const ClearPushWSSubscriptions = () => {
   }, []);
   return null;
 };
+
+const eventEmitter =
+  NativeModules?.StreamVideoReactNative &&
+  new NativeEventEmitter(NativeModules?.StreamVideoReactNative);
```
Comment (oliverlaz) [inline packages/react-native-sdk/src/providers/StreamCall.tsx:159]:
```suggestion
const eventEmitter =
  NativeModules?.StreamVideoReactNative 
    ? new NativeEventEmitter(NativeModules?.StreamVideoReactNative)
    : undefined;
```

Let's be a bit more explicit.

**Review by oliverlaz — APPROVED** [review]
